### PR TITLE
Add DodoSource

### PR DESCRIPTION
**Pull request by nrbgt**
## References

- continues #44
- continues #7

## Code changes

- [x] add some more demo deps
  - [x] `ipydatagrid`
  - [x] ~~`ipytree`~~ 
    - [x] already quite busy, wouldn't add much
  - [x] `ipylab`   
  - [x] updated `ipywidgets` and `jupyterlab_widgets`
    - [x] update serialization for strict JSON
- [x] add `DodoSource`
  - [x] find nodes and lnks
    - [x] tasks
    - [x] files
    - [x] subtasks 
    - [x] task dep
    - [x] folders-of-files
- [ ] example notebook 
  - [x] make panel
  - [x] load source
  - [x] make grids
  - [x] link grids and selection
  - [x] run tasks
    - [x] logs without killing UI
    - [x] without blocking kernel
  - [x] highlight 
    - [x] dependency paths following `file_dep` and `task_dep` links
    - [x] upstreams
    - [x] downstreams 
  - [x] collapse nodes and edges 
- [x] issues noticed while working on app
  - [x] handle in-place updates (for certain cases)
  - [x] handle `linkHandlers` and `nodeHandlers` better for facets
  - [x] fix multiple shapes with "canary" attributes (e.g. `Rectangle.width` and `Text.text`)
  - [x] add `line_dash` to more shapes (via HasStrokeAndFill) 
  - [x] fix center point of custom Rectangle and Ellipse shapes 
- [x] docs
  - [x] changelog
  - [x] clean up some hard-coded positions for facets in docs
  - [x] update `jupyterlite-core` and `pyodide-kernel`

## User-facing changes

- users of the binder demo will be able to try out the grid against an evolving system
- users of the lit demo will see:

  | note | screen | 
  |-|-|
  | custom shapes | image
  | running tasks based on task selection | image
  | ipylab icons | image
  | shortest path with particles |image
  | upstream/downstream highlight | image
 

## Backwards-incompatible changes

- n/a

### CODE REVIEW

**Review by sanbales — COMMENTED** [review]
Looking good, just some small things I ran into

File: examples/DodoSource.ipynb [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b134dfda-6517-4734-b59a-beac6b8c714a",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "app = JupyterFrontEnd()\n",
+    "panel = Panel([])\n",
```
Comment (sanbales) [inline examples/DodoSource.ipynb:63]:
```suggestion
    "panel = SplitPanel([], layout=dict(height="100%"))\n",
```

File: examples/DodoSource.ipynb [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "grid_opts = dict(\n",
+    "    selection_mode=\"row\",\n",
+    "    layout=dict(height=\"100%\", max_height=\"100%\", min_height=\"400px\"),\n",
+    ")\n",
+    "n_grid = DataGrid(fg.source.nodes, **grid_opts)\n",
+    "l_grid = DataGrid(fg.source.links, **grid_opts)\n",
+    "panel.children = [fg, W.HBox([n_grid, l_grid])]"
```
Comment (sanbales) [inline examples/DodoSource.ipynb:164]:
```suggestion
    "panel.children = [fg, SplitPanel([n_grid, l_grid], orientation=\"horizontal\")]"
```

File: src/ipyforcegraph/sources/dodo.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    def _default_deps(self) -> Dependency:
+        return Dependency(
+            self.BACKENDS[self.backend], str(self.project_root / ".doit.db")
+        )
+
+    @T.validate("project_root")
+    def _validate_project_root(self, proposal: Any) -> Path:
+        return Path(proposal.value).resolve()
+
+    def _reload_tasks(self) -> List[Task]:
```
[the remark this answers is not in the document]
Comment (nrbgt) [reply src/ipyforcegraph/sources/dodo.py]: a4f7c75

File: examples/DodoSource.ipynb [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ccd5095a-e5f3-41e5-b121-e7b866e5376e",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "link_grid_and_graph(n_grid, n_sel)\n",
+    "link_grid_and_graph(l_grid, l_sel)"
```
[the remark this answers is not in the document]
Comment (nrbgt) [reply examples/DodoSource.ipynb]: a4f7c75

File: examples/DodoSource.ipynb (hunk 1 above)
Comment (nrbgt) [reply examples/DodoSource.ipynb]: a4f7c75

File: examples/DodoSource.ipynb [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a9a3bc71-decf-4ec2-b351-4cf1bdaf7af0",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "source = DodoSource(project_root=\"..\")\n",
+    "T.dlink((source, \"project_root\"), (panel.title, \"label\"), lambda x: x.name);"
```
[the remark this answers is not in the document]
Comment (nrbgt) [reply examples/DodoSource.ipynb]: a4f7c75

File: examples/DodoSource.ipynb (hunk 2 above)
Comment (nrbgt) [reply examples/DodoSource.ipynb]: a4f7c75

File: js/widgets/serializers/widget.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2023 ipyforcegraph contributors.
+ * Distributed under the terms of the Modified BSD License.
+ */
+import { WidgetModel, unpack_models } from '@jupyter-widgets/base';
+
+export const widget_serialization = {
+  deserialize: unpack_models,
+  serialize: (models: WidgetModel[]): string[] => {
```
Comment (nrbgt) [inline js/widgets/serializers/widget.ts:9]: this addresses the issue (also seen elsewhere) around the most recent `@jupyter-widgets/base` being more strict about its serialization guarantees. Might get added upstream, but :shrug:.

**Comment by nrbgt** [comment]
Resolve with `networkx` incoming...

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,10 +6,17 @@
...
 - improves default selection behavior for `LinkShapes.curvature` and
   `LinkShapes.line_dash`
+- adds `line_dash` to all 2D shapes
 - removes the significance of order in `ForceGraph.behaviors`
   - all node, link, and graph behaviors now have a (sensible default) `.rank` trait
     which determines the order in which they are applied.
   - lower `rank` are applied before higher `rank`
+- adds `DodoSource` for interpreting `doit` tasks graphs
+- adds `node_preserve_columns`, `link_id_column`, and `link_preserve_columns`
+  - these allow for keeping values when updating data, such as those created by the
```
Comment (sanbales) [inline CHANGELOG.md:16]: Just wanted to double check this was still going in this PR...

**Review by sanbales — APPROVED** [review]
Looks good to me, just a couple of small comments, but happy to merge if you are cool with it.

File: docs/_static/logo.svg
[the remark this answers is not in the document]
Comment (nrbgt) [reply docs/_static/logo.svg:1]: it has just been compressed.

File: CHANGELOG.md (hunk 7 above)
Comment (nrbgt) [reply CHANGELOG.md:16]: yep, it's in here, and the demo shows _some_ improvement in the stability of the render, though the vanilla forces still encourage stuff to move

File: src/ipyforcegraph/behaviors/shapes.py [hunk 8]
```diff
@@ -23,7 +23,10 @@
 
 @W.register
 class Text(HasFillAndStroke):
-    """Draw a text shape, with an optional background."""
+    """Draw a text shape, with an optional background.
+
+    If the ``text`` trait is (or evaluates to) ``0`` or ``None``, no shape will be drawn.
```
Comment (sanbales) [inline src/ipyforcegraph/behaviors/shapes.py:28]: Would the same happen with ``height``?

**Review by sanbales — APPROVED** [review]
Ran a `doit lab` locally and played with the `DodoApp.ipynb` for a while.

Seems to be working well.  Only things I've noticed (and these may be PEBCAK or not intended):
1. selection from `datagrid` to graph does not seem to work (i.e., selecting a row on the grids, does not highlight `nodes` or `links` on the graph)
2. sometimes the between particles do not seem to appear for things someone might expect (e.g., from `env` to `setup`)

Tried it on `lite` as well but didn't see the `Dodo` notebooks in there.

**Comment by nrbgt** [comment]
thanks for the review! :heart_eyes: 

> didn't see the `Dodo` notebooks in there.

Yep, they don't work in lite because `subprocess.Popen`. Also `doit` doesn't import, due to some unvendored stdlib packages... to that end, I started some playground stuff on [doitoml](https://doitoml.readthedocs.io/en/latest/demo.html), but if course _doit_ relies on `subprocess.Popen` as well, so only pure-python actions work.

> the between particles

will look into it. Trying to get fancy with non-blocking stuff. Got some ideas on how that could be improved.

> selection from datagrid

will look into it

**Comment by nrbgt** [comment]
>> selection from datagrid
>
> will look into it

oof, this is general failure, even setting it programmatically doesn't work.

```
node_selection.selected = [0]
# doesn't highlight `all`
```

the issue is that we don't update the `graphData.nodes._selected` when a selection update comes in, and then `node-selection` thing is getting clobbered by the fancy text junk.

**Comment by nrbgt** [comment]
fixed selection on 43202f1

File: src/ipyforcegraph/behaviors/shapes.py (hunk 8 above)
Comment (nrbgt) [reply src/ipyforcegraph/behaviors/shapes.py:28]: it would _visually_ hide it, but you'd still pay the canvas price

**Comment by nrbgt** [comment]
>> the between particles
> will look into it. 

Ah, this might due to it actually being a bit more accurate now, as it will only show particles if there is a path of the form:

```
[file] --> file_dep --> task --> targets --> [other file]
```

(Or the reverse)

Initially, it kinda lied, as it would follow sibling paths like:

```
[file] --> file_dep --> task --> file_dep [other file]
```

So basically: a lot of the nodes aren't particularly sensitive to other nodes. If there's a concrete example, happy to look into it!

**Comment by nrbgt** [comment]
I've updated the label to show the path count, a la image

**Comment by sanbales** [comment]
> > > the between particles
> > > will look into it.
> 
> Ah, this might due to it actually being a bit more accurate now, as it will only show particles if there is a path of the form:
> 
> ```
> [file] --> file_dep --> task --> targets --> [other file]
> ```
> 
> (Or the reverse)
> 
> Initially, it kinda lied, as it would follow sibling paths like:
> 
> ```
> [file] --> file_dep --> task --> file_dep --> [other file]
> ```
> 
> So basically: a lot of the nodes aren't particularly sensitive to other nodes. If there's a concrete example, happy to look into it!

Yeah, that makes sense.  No, I don't think it's worth it to go beyond what you've done.  It's working nicely!

**Comment by sanbales** [comment]
I vote we pull the trigger.  The only thing I could find was that when you try to open some files, it says it can't find them.  Think it is for files that start with `.`?  Obviously, testing this on 🪟ers, so 🐧 may be fine...

image

**Pull request merged**